### ISSUE TITLE
Integration with Angular2 / TypeScript

### ISSUE THREAD

**Issue by flyingmutant** [rollup/rollup#280]
Rollup seems like a dream bundler for ng2 apps, but it is unclear how it can be integrated with typescript codebases.

What is the recommended way to rollup a typescript app, which uses es6 imports, and which depends on ng2 (which is also written in typescript and uses es6 imports exclusively)?

**Comment by Victorystick** [comment]
The recommended way is to use (the not-yet-created) [`rollup-plugin-typescript`](https://github.com/rollup/rollup-plugin-typescript) in a `rollup.config.js` file like so:

```js
import typescript from 'rollup-plugin-typescript';

export default {
  entry: 'src/app.ts',
  dest: 'dist/app.js',

  format: 'iife',

  plugins: [
    typescript()
  ]
};
```

Take a look at [`rollup-plugin-npm`](https://github.com/rollup/rollup-plugin-npm) and [`rollup-plugin-babel`](https://github.com/rollup/rollup-plugin-babel) (or why not Rollup's own [`rollup.config.js`](https://github.com/rollup/rollup/blob/master/rollup.config.js) file) to get a feeling for how it should eventually all fit together.

**Comment by robwormald** [comment]
tracking this as well. really impressed with rollup so far.

From the angular2 end, there's no reason we can't ship an ES6 build (we already do, though we're in the process of removing it...), though with a rollup-plugin-typescript I suppose we could just use the TS source. 

It would require a bit of refactoring on our end wrt to external paths, as we use a lot of `import {Thing} from 'angular2/src/foo/bar' ` type paths. 

Definitely interested in digging into this. @Victorystick would you mind having a chat with me and @pkozlowski-opensource at some point in the near future? Rollup would solve a ton of issues on our end (I'd love to have it built into our CLI tool!)

**Comment by Victorystick** [comment]
@robwormald Absolutely! And actually, you can create a custom plugin that only handles the external `angular2` paths so that you needn't refactor. It shouldn't be more than a few lines long.

**Comment by Victorystick** [comment]
Something like this would do:

```js
const path = require( 'path' );

const angularRoot = __dirname;
const angular2 = 'angular2/';

module.exports = {
  entry: './app.ts',

  plugins: [
    typescript(),

    {
      // ignore everything that doesn't start with 'angular2/'
      // resolve the rest within the angular directory
      resolveExternal: function ( id ) {
        if ( !id.startsWith( angular2 )) return null;

        return path.join( angularRoot, id.slice( angular2.length ) );
      }
    }
  ]
};
```

**Comment by robwormald** [comment]
@Victorystick @Rich-Harris  wow, I got that working with only a tiny bit of fiddling from our ES6 build. color me *impressed*

Would you mind opening up a rollup-plugin-typescript repo? I'm hot to trot now :D

**Comment by Rich-Harris** [comment]
I've created that repo – just having a little trouble getting TypeScript to emit ES6 modules! Do you know what the necessary options are? https://github.com/rollup/rollup-plugin-typescript/issues/1

**Comment by Victorystick** [comment]
@flyingmutant Did you get something working with [rollup/rollup-plugin-typescript](https://github.com/rollup/rollup-plugin-typescript)?

**Comment by flyingmutant** [comment]
Just tried it. After changing all my app imports from `import './thing'` to `import './thing.ts'` (is there a way to avoid it?), I managed it to work, except for

```
Treating 'angular2/angular2' as external dependency
Treating 'angular2/http' as external dependency
Treating 'angular2/router' as external dependency
Treating '@reactivex/rxjs/dist/cjs/Rx' as external dependency
```

All of these live under `node_modules` and tsc picks them automatically just fine. How can I tell rollup to bundle them as well?

**Comment by Victorystick** [comment]
Are you using [the `npm` plugin](https://github.com/rollup/rollup-plugin-npm)? It is required for Rollup to resolve imports from `node_modules`.

**Comment by flyingmutant** [comment]
Also, typescript run by rollup seem to ignore any errors (that tsc will report) – is this intentional?

**Comment by Victorystick** [comment]
No, that is not intentional. Transpilation errors should be reported. It's just early days for the TypeScript plugin. Please [create an issue](https://github.com/rollup/rollup-plugin-typescript/issues), and I'll see what I can do.

**Comment by robwormald** [comment]
It's going to come down to whether or not @Victorystick wants to support basic transplilation, or a full fledged TypeScript backend. I would definitely prefer the latter, but that's a significantly more complicated task to implement (and it would preferable to implement such a plugin *in* TypeScript, imho) 

TS provides a [Langugage Services API](https://github.com/Microsoft/TypeScript/wiki/Using-the-Language-Service-API) and a [Compiler API](https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API) that would basically need to be hooked into rollup-plugin-ts (and I suspect a few tweaks might be needed at the rollup end WRT plugin APIs)

As long as @Victorystick is OK with it, i'd like to make this a pure TS project - (working against the Typescript API from JS is no fun) and move forward on implementing the compiler backend.

**Comment by robwormald** [comment]
WRT the npm plugin, TS has node resolution built in now, and it may be better to bake it in to this plugin - packages like angular2 and @reactivex distribute files like, and you want the TS compiler to be able to understand both (or optionally opt into using the original TS source, I guess?) 
```
angular2.js
angular2.d.ts
```
Typescript's recent change to adding a `typings: "index.d.ts"` key to package.json files makes this pretty easy to work with.

**Comment by flyingmutant** [comment]
With `rollup-plugin-npm`, I get stuff like 

```
Export 'bootstrap' is not defined by '/Users/gp/.../node_modules/angular2/angular2.js'
```

For 

```typescript
import  * as ng from 'angular2/angular2';
ng.bootstrap();
```

(looks like rollup simply appends the `.js` to get the module file). Trying

```typescript
import * as ng from 'angular2/es6/prod/angular2';
```

instead gives

```
Error: Module /Users/gp/.../node_modules/angular2/src/facade/lang.js does not export RegExpWrapper (imported by /Users/gp/.../node_modules/angular2/es6/prod/src/common/pipes/number_pipe.js)
```

because `number_pipe.js` imports `''angular2/src/facade/lang'` = `'angular2/src/facade/lang.js'`, which is not in ES6.

**Comment by Victorystick** [comment]
There's a PR rollup/rollup-plugin-npm#3 in progress to use Node's resolve algorithm instead of the current one. I'm not sure whether this will work as expected for TypeScript code, so it might be best to let the TypeScript plugin handle the resolution as well.

If it makes development easier to use TypeScript for development of the plugin, then let's go ahead an do that. :+1:

**Comment by cetra3** [comment]
Now that Angular 2.0 beta is out, it'd be nice if there was no friction in using rollup.js with Angular 2.

It would be good to have a working rollupjs config/tutorial compatible with with the [Quickstart](https://angular.io/docs/ts/latest/quickstart.html).  Judging from this issue though, rollup still needs a bit of tweaking?  What needs to be done?

**Comment by eventualbuddha** [comment]
@cetra3 I'm guessing rollup itself doesn't need much changing, but the TypeScript plugin may. I'm not very familiar with either Angular 2 or TypeScript, so I'm not going to know exactly what is needed to improve support. Has anyone tried this integration recently?

**Comment by cetra3** [comment]
@eventualbuddha There are 3 things I can think of off the top of my head.

### rollup.js
Need some documentation around how to address the dreaded `Treating blah as external dependency`.  I have no idea what I'm meant to do to fix it.  I'd like to be able to let rollup js know where it can find these in an easier fashion.  Whether this would be to include the `rollup-plugin-npm` or create a `resolveExternal` function.

### rollup-plugin-typescript:  It only resolves the *first* typescript file, not imports.  If you have a statement like:
```
import {AppComponent} from './app.component'
```
And app.component happens to be the file `app.component.ts`, then rollup won't be able to find this file either.  I have a feeling that this bug has not been hit before, as the default typescript action is to place a js file alongside of it.

### Example project
While probably out of scope of rollup.js, we really need an example project that will show off the angular 2 quickstart with rollup.js, including all the bits and pieces needed for it to work.  I haven't been able to get a working project with rollup yet, which bundles the quickstart app, alongside all the angular 2 bits in a single bundle, so I don't even know if it's possible at this stage.

**Comment by eventualbuddha** [comment]
@cetra3 It seems like [the part of rollup that adds `.js` extensions](https://github.com/rollup/rollup/blob/b33f14f2b0bdb8fc2d2a74bc830e366766452e0e/src/utils/defaults.js#L8-L28) could be improved to take a list of extensions which could be provided by the plugins. Alternatively, you could just specify the extension.

Unfortunately there seems to be a bug in the current version of TypeScript that makes targeting ES6 modules fail. Transforming `export class App {};` yields this:

```js
var App = (function() {
  function App() {
  }
  return App;
})();
App = App;
```

As for the example project, you could base it off [`rollup-starter-project`](https://github.com/rollup/rollup-starter-project).

**Comment by cetra3** [comment]
@eventualbuddha My gut feel is that we should use typescript's `nodeModuleNameResolver` function and use it within a `resolveId` function.  I'm having a look at this now to see what needs to be done.  The most suitable place for this would be in the typescript plugin.

**Comment by eventualbuddha** [comment]
@cetra3 Sounds great! Would love to see a PR.

**Comment by cetra3** [comment]
@eventualbuddha I have published a PR for resolving imports using typescript, but some tests will need to be rewritten. I don't feel that the typescript plugin should be responsible for transpiling to ES5, this should be the responsibility of babel or others, so the tests will need to be rewritten.  Pull request is here:
https://github.com/rollup/rollup-plugin-typescript/pull/15

As for sorting out *external dependencies* this is a hard one, and one I don't know I can answer easily.  The issue is that some typescript libraries (angular 2 included) have `.d.ts` files which act like interfaces, but don't contain any code implementation.  These files are used so that typescript can check type information with external libraries, such as jQuery etc...  But they aren't useful when you're trying to package everything in a bundle.  

There are a couple of paths to follow here:

* Ignore any external dependencies when transpiling, and handle dependencies with another plugin. This is the current behaviour of the pull request and seems to be most sensible in the light of typings.  Unfortunately this doesn't work for angular 2 within the npm package at the moment, as angular appears to be a blend of ES6 and CommonJS modules...

* Allow the typescript module to search for dependencies but ignore the `.d.ts` typings.  This appears to have more robustness, but you will still need to use npm and commonjs.  I've managed to get this sort of to work and bundle a package, but a few of the dependencies are not included, due to some funkiness in Angular 2, and the browser is getting a few undefined errors.

**Comment by robwormald** [comment]
>  I don't feel that the typescript plugin should be responsible for transpiling to ES5
Why not? Why should a typescript user (who may rely on Typescript specific functionality) need to provide another transpiler (which may or may not provide such features - see: decorators, metadata) - that doesn't make a whole lot of sense to me.

> The issue is that some typescript libraries (angular 2 included) have .d.ts files which act like interfaces, but don't contain any code implementation. 

Going forward the vast majority of newly authored Typescript modules will do this, fwiw - this is to aid ease of use (so that consumers of an NPM package don't need to re-transpile their code). This sort of precludes rollup usage (though its effectively the same issue current NPM packages would have and the reason for the `jsnext:main` idea.

> Unfortunately this doesn't work for angular 2 within the npm package at the moment, as angular appears to be a blend of ES6 and CommonJS modules...

Angular2 is currently shipped to NPM with CJS at the root, and ES6/TS source in subdirectories. It's authored in 100% TS (and thus that source could be used for rollup purposes)

---

The issue here is that the TS compiler provides pretty much all the functionality a TS consumer needs - transpilation, type checking, etc, *except* bundling. I think the better solution is going to be relying on *more* of Typescript's built in functionality, not less - introducing Babel into the mix is a non-starter.

**Comment by cetra3** [comment]
>Why not? Why should a typescript user (who may rely on Typescript specific functionality) need to provide another transpiler (which may or may not provide such features - see: decorators, metadata) - that doesn't make a whole lot of sense to me.

There are two good reasons why I have changed this in the PR: 
* Rollup.js is designed for ES6 modules, so you need to target ES6.  If you don't, then you can't use more than one file. 
I have tried leaving it with ES5,  I'm using the [angular tutorial](https://angular.io/docs/ts/latest/tutorial/toh-pt2.html) to test, and I get this error:

```
Module /Users/cetra/git/ng2test/app/app.component.ts does not export AppComponent (imported by /Users/cetra/git/ng2test/app/boot.ts)
```
Leaving the target at ES6 doesn't break the tutorial. You could possibly target ES5 then use the commonjs plugin to bring it back up to ES6 but that is just a bit silly.

* The current configuration silently breaks anyway, it's just not reported by the plugin.  This is the error that comes back when setting `reportDiagnostics : true` is:
```
{ file: undefined,
    start: undefined,
    length: undefined,
    messageText: 'Cannot compile modules into \'es6\' when targeting \'ES5\' or lower.',
    category: 1,
    code: 1204
}
```

I agree that typescript compiler options would be a good thing to expose though, easiest would be to allow the user to change the typescript compiler options in rollup.config.js

**Comment by robwormald** [comment]
Let's do this as it *should* be done, not based on an arbitrary set of instructions from the tutorial! I will *personally* merge the PR that changes the tutorial to use rollup when such a thing is available ;-)

Angular2 is a ridiculously complex test case - I suggest we start with something slightly simpler - two ES6 style typescript files using two npm installed packages - one that distributes `.d.ts` files alongside, and one that doesn't.

**Comment by cetra3** [comment]
Ok, sounds good.  I have a fair way of being able to get that to work so I'll set up an example repo.

**Comment by Victorystick** [comment]
> The issue here is that the TS compiler provides pretty much all the functionality a TS consumer needs - transpilation, type checking, etc, _except_ bundling. I think the better solution is going to be relying on more of Typescript's built in functionality, not less [...]

@robwormald I completely agree. I've raised some of the issues I encountered during by (brief) work on Rollup's TypeScript plugin, see Microsoft/TypeScript#5790.

@cetra3 It makes most sense for TypeScript's compiler to handle transpilation, and (I believe) most people will want to generate ES5 that's usable today, instead of ES6 all that's supported by a single runtime. Rollup requires JavaScript that uses _ES modules_, not necessarily all of ES2015. That's why the plugin is configured as it is. It's unfortunate that the compiler doesn't support proper ES modules when targeting ES5.

I too want to make bundling of Angular 2 apps possible.

**Comment by cetra3** [comment]
OK @robwormald I have created a simple rollup test repo here: https://github.com/cetra3/rollup-ts-test which works after I have made the [following change](https://github.com/cetra3/rollup-plugin-typescript/commit/9649386f10798ae3f9a1d2b44efcec12aa8adced) to the rollup plugin.

It's a very trivial implementation of trying to find the appropriate js file (replaces `d.ts` with `.js` and checks if exists).   But it works in the artificial case.  I still use babel to compile it down to ES5, as I can't see a way of getting around this currently with TypeScript compiler not supporting blended ES5/ES6.

I feel like there needs to be a formal way of correlating `d.ts` files with `.js` files, because at the moment everyone is doing it differently and the [wiki](https://github.com/Microsoft/TypeScript/wiki/Typings-for-npm-packages) does not cover all uses cases (i.e, what if you have multiple files/exports instead of just one?)

I have also created an Angular 2 rollup repo, which is based off the heroes tutorial (which isn't working at the moment, obviously): https://github.com/cetra3/rollup-ng2-test.

To get this working, angular2 and friends should have seperate packages that are either pure typescript, or ES module packages with `.d.ts` files in the right place so rollup + plugins can do their thing.

If it's a pure typescript module, it should just work with the plugin currently. 

If it's an ES module with typings, we should not need to rely on the commonjs plugin (which I currently have, but would rather not have, in the test repo). 

Making an angular2 typescript only package in npm would be a good start!

**Comment by Victorystick** [comment]
I thought I'd give an update about the state of the TypeScript plugin. I released version `0.4.0` a few days ago.

@flyingmutant The plugin now handles imports without the `.ts` extension

> Unfortunately there seems to be a bug in the current version of TypeScript that makes targeting ES6 modules fail. Transforming `export class App {};` yields this:
> ```js
> var App = (function() {
  function App() {
  }
  return App;
})();
App = App;
```

@eventualbuddha Thanks to [a small hack](https://github.com/rollup/rollup-plugin-typescript/blob/master/src/fixExportClass.ts), the plugin now generates code like this:
```js
var App = (function() {
  function App() {
  }
  return App;
})();
export { App };
```

@cetra3 @robwormald The plugin can bundle an Angular2 _app_, but requires Angular to be marked as an external (https://github.com/cetra3/rollup-ng2-test/pull/1). We'll see how to go about bundling Angular2 itself going forward.

**Comment by arturparkhisenko** [comment]
Just wanted create es2015 app using Angular2 and got this problem, now waiting babel6 + annotations support:
```js
//code before..
@Component({
  selector: 'my-app',
  template: templateHelloWorld
})
```
Error: `SyntaxError: Unexpected character '@'`
Babel plugin: https://github.com/shuhei/babel-plugin-angular2-annotations

**Comment by robwormald** [comment]
@Victorystick nice work :+1:

**Comment by cetra3** [comment]
The new `--allowJS` option of typescript that [landed](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#including-js-files-with---allowjs) in `1.8` may be of use here for getting Angular2 bundled

**Comment by Victorystick** [comment]
@arturparkhisenko Have you tried [rollup-plugin-babel](https://github.com/rollup/rollup-plugin-babel)? You can use any Babel transforms you need. :smiley:

**Comment by arturparkhisenko** [comment]
@Victorystick maybe or i need to use this packages? as i see rollup-plugin-babel is using them:
```json
{
  "dependencies": {
    "babel-preset-es2015": "^6.3.13",
    "babel-plugin-angular2-annotations": "^5.0.0",
    "babel-plugin-transform-class-properties": "^6.3.13",
    "babel-plugin-transform-decorators-legacy": "^1.3.3",
    "babel-plugin-transform-flow-strip-types": "^6.3.15"
  }
}
```

**Comment by TrySound** [comment]
@Victorystick  Can be closed?

**Comment by drejohnson** [comment]
Any update on getting Angular 2 working with Rollup

**Comment by pkozlowski-opensource** [comment]
@drejohnson see https://github.com/rollup/rollup-plugin-typescript/issues/9#issuecomment-175679272

tl;dr; you can bundle an Angular2 app with rollup with some tweaks here and there. But if you using TypeScript the gains will be minimal...

**Comment by drejohnson** [comment]
@pkozlowski-opensource Okay Thanks!

**Comment by TrySound** [comment]
Looks like resolved.

**Comment by cetra3** [comment]
I don't really think it is resolved.  There is still a lot of work to be done for this. Unless you feel that Angular itself should not be capable of being rolled up?  I.e, always an external dependency

**Comment by cetra3** [comment]
Hi guys,

Just after an update for this.  I know that @robwormald has made some [progress](https://news.ycombinator.com/item?id=11618024) but not sure what state it's in since then.

**Comment by ffchris** [comment]
I've been getting errors like "toPromise.js does not export toPromose (imported by ... async.js)" when I try bundling angular2. I'm wondering is this is related to the discussion here.

**Comment by ffchris** [comment]
basically, this results in the "does not export" error:
https://medium.com/@jonnysamps/angular-2-bundling-with-rollup-4738d0148a2c

**Comment by jkuri** [comment]
I managed to get this working, you can see example at https://github.com/jkuri/angular2-rollup-starter.

**Comment by Shyam-Chen** [comment]
Tree-shaking with Rollup and Angular 2.

https://github.com/Shyam-Chen/Angular2-Rollup-Demo

**Comment by sylvainpolletvillard** [comment]
I am in the same situation than @arturparkhisenko . I have an Angular2 application that builds with Babel and this plugin: https://github.com/shuhei/babel-plugin-angular2-annotations . The bundle size is gigantic (> 1mb only for Angular sources) since there is no tree shaking at all ; so rollup would be a life saver here.

The problem is these Angular 2 annotations.
If I don't use rollup-plugin-babel, I got `SyntaxError: Unexpected character '@'`
If I do use rollup-plugin-babel, I got `Unexpected token import`

Would it be possible for Rollup to have an option that just ignore these annotations ?

**Comment by Rich-Harris** [comment]
Am doing some tidying up and seeing which old issues can be closed. This thread has got a bit unwieldy and it's not quite clear what outstanding requirements there are, so I'm going to close it — if there are things that are still not behaving as expected could people reopen those as self-contained issues? Thanks
